### PR TITLE
Test app directory

### PR DESCRIPTION
**Pull request by lealem47**
Added test_apps apps directory

### CODE REVIEW

**Review by dgarske — CHANGES_REQUESTED** [review]
1) We still need `examples/server/server` and `examples/client/client`, but they should be a stripped down simple version of a TLS client / server most of the arguments still. Remove the benchmarking code, default to HTTP, remove all the -H internal test arg basically anything that is testing specific.
2) Do not move the `examples/config` or `examples/benchmark` directories. Possibly don't move the echo examples either.
3) You missed updating the `./scripts/benchmark.test` file.

Also please talk to someone about how to do a squash merge and rebase.

File: test_apps/configs/README.md [hunk 1]
```diff
@@ -19,4 +19,4 @@ Example wolfSSL configuration file templates for use when autoconf is not availa
 To use these with autoconf:
 
 1. Copy file to root as `user_settings.h`.
-2. Run `./configure --enable-usersettings --disable-examples && make`
+2. Run `./configure --enable-usersettings --disable-test_apps && make`
```
Comment (elms) [inline test_apps/configs/README.md:22]: If we are changing the configure option I don't see any of the changes in `configure.ac`. I'm not sure if we want to change that or not....it may break peoples build when they upgrade.

Comment (dgarske) [reply test_apps/configs/README.md:22]: Agree, that should not change and doc find/replace needs to be reverted.

Comment (lealem47) [reply test_apps/configs/README.md:22]: All docs have been reverted and changed so that only directories are modified

Comment (lealem47) [reply test_apps/configs/README.md:22]: *directories and file paths

File: CMakeLists.txt [hunk 2]
```diff
@@ -1222,7 +1222,7 @@ endif()
 
 set(WOLFSSL_EXAMPLES_HELP_STRING "Enable examples (default: enabled)")
 add_option("WOLFSSL_EXAMPLES" ${WOLFSSL_EXAMPLES_HELP_STRING} ${EXAMPLES_DEFAULT} "yes;no")
-
+add_option("WOLFSSL_EXAMPLES" ${WOLFSSL_EXAMPLES_HELP_STRING} ${EXAMPLES_DEFAULT} "yes;no")
```
Comment (dgarske) [inline CMakeLists.txt:1225]: This line is now duplicated. Guessing it should be for the test_apps?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1519,8 +1535,8 @@ if(WOLFSSL_EXAMPLES)
         tests/srp.c
         tests/suites.c
         tests/unit.c
-        examples/server/server.c
-        examples/client/client.c)
+        test_apps/server/test_server.c
```
Comment (dgarske) [inline CMakeLists.txt:1538]: It should still build the original examples/server/server.c and examples/client/client.c

File: IDE/ECLIPSE/DEOS/README.md [hunk 4]
```diff
@@ -41,7 +41,7 @@ Note: To work wolfssl directory must not be under the workspace directory.
 ### Setting up a Deos project with wolfSSL
 
 The `tls_wolfssl.c` example application provides a simple function to
-run the selected examples at compile time through the following four
+run the selected test_apps at compile time through the following four
```
Comment (dgarske) [inline IDE/ECLIPSE/DEOS/README.md:44]: Please review these changes in this file. We still point people to the same original example.

File: IDE/ECLIPSE/DEOS/tls_wolfssl.c [hunk 5]
```diff
@@ -455,7 +455,7 @@ void wolfssl_server_test(uintData_t statusPtr)
     }
     /* accept client socket connections */
     printf("Listening for client connection\n");
-    printf("E.g, you can use ./examples/client/client.exe -h 192.168.219.100\n");
+    printf("E.g, you can use ./test_apps/client/test_client.exe -h 192.168.219.100\n");
```
Comment (dgarske) [inline IDE/ECLIPSE/DEOS/tls_wolfssl.c:458]: Revert this change.

File: IDE/ECLIPSE/MICRIUM/README.md [hunk 6]
```diff
@@ -9,7 +9,7 @@ You can start with your IDE-based example project for Micrium uC/OS-III and uC/T
 
 wolfSSL supports a compile-time user configurable options in the `IDE/ECLIPSE/MICRIUM/user_settings.h` file.
 
-The `wolfsslRunTests.c` example application provides a simple function to run the selected examples at compile time through the following four #defines in user_settings.h.
+The `wolfsslRunTests.c` example application provides a simple function to run the selected test_apps at compile time through the following four #defines in user_settings.h.
```
Comment (dgarske) [inline IDE/ECLIPSE/MICRIUM/README.md:12]: Revert all changes in this file.

File: IDE/Espressif/ESP-IDF/examples/wolfssl_client/README.md [hunk 7]
```diff
@@ -15,8 +15,8 @@ The Example contains of wolfSSL tls client demo.
 When you want to test the wolfSSL client
 
 1. "make flash monitor" to load the firmware and see the context  
-2. You can use <wolfssl>/examples/server/server program for test.  
+2. You can use <wolfssl>/test_apps/server/test_server program for test.
```
Comment (dgarske) [inline IDE/Espressif/ESP-IDF/examples/wolfssl_client/README.md:18]: Revert all changes in this file.

File: IDE/MDK-ARM/Projects/MDK-ARM-STM32F2xx.uvopt [hunk 8]
```diff
@@ -569,7 +569,7 @@
       <Focus>0</Focus>
       <tvExpOptDlg>0</tvExpOptDlg>
       <bDave2>0</bDave2>
-      <PathWithFileName>..\..\..\examples\client\client.c</PathWithFileName>
+      <PathWithFileName>..\..\..\test_apps\client\client.c</PathWithFileName>
```
Comment (dgarske) [inline IDE/MDK-ARM/Projects/MDK-ARM-STM32F2xx.uvopt:572]: Revert all changes in this file.

File: IDE/RISCV/SIFIVE-UNLEASHED/README.md [hunk 9]
```diff
@@ -83,8 +83,8 @@ Copy files to device (replace IP address):
 ```sh
 scp ./wolfcrypt/test/testwolfcrypt root@192.168.0.144:~
 scp ./wolfcrypt/benchmark/benchmark root@192.168.0.144:~
-scp ./examples/client/client root@192.168.0.144:~
-scp ./examples/server/server root@192.168.0.144:~
+scp ./test_apps/client/test_client root@192.168.0.144:~
```
Comment (dgarske) [inline IDE/RISCV/SIFIVE-UNLEASHED/README.md:86]: Revert all changes in this file.

**Review by dgarske — CHANGES_REQUESTED** [review]
Please revert the changes in IDE examples. This needs to be mostly transparent. Basically taking the existing complex examples/server/server.c and client and moving them to test_apps. The existing examples stay but get trimmed down. No benchmarking, or test specific code. Also please talk to someone about how to rebase and squash merge...

File: CMakeLists.txt (hunk 3 above)
Comment (lealem47) [reply CMakeLists.txt:1538]: I kept this change so that unit.test would pass

File: test_apps/server/test_server-ntru.vcproj [hunk 10]
```diff
@@ -0,0 +1,199 @@
+<?xml version="1.0" encoding="Windows-1252"?>
```
Comment (dgarske) [inline test_apps/server/test_server-ntru.vcproj:1]: Go ahead and remove the NTRU examples from the test_apps.

File: test_apps/server/test_server.vcproj [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+	<References>
+	</References>
+	<Files>
+		<Filter
+			Name="Source Files"
+			Filter="cpp;c;cc;cxx;def;odl;idl;hpj;bat;asm;asmx"
+			UniqueIdentifier="{4FC737F1-C7A5-4376-A066-2A32D752A2FF}"
+			>
+			<File
+				RelativePath=".\server.c"
```
Comment (dgarske) [inline test_apps/server/test_server.vcproj:178]: This will need to be updated to test_server.c

File: CMakeLists.txt [hunk 12]
```diff
@@ -1593,8 +1609,10 @@ set(INSTALLED_EXAMPLES
     ${CMAKE_CURRENT_SOURCE_DIR}/examples/sctp/sctp-client.c
     ${CMAKE_CURRENT_SOURCE_DIR}/examples/sctp/sctp-server-dtls.c
     ${CMAKE_CURRENT_SOURCE_DIR}/examples/echoclient/echoclient.c
-    ${CMAKE_CURRENT_SOURCE_DIR}/examples/server/server.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/test_apps/server/test_server.c
```
Comment (dgarske) [inline CMakeLists.txt:1612]: Please group these... test_ next to each other. Also add a `# comment` for why examples/client and server are not in `add_executable` above (because they are already covered here). And confirm this with building or @haydenroche5

**Review by dgarske — CHANGES_REQUESTED** [review]
This is looking very good. Just a couple very minor items and it should be ready.

**Review by dgarske — APPROVED** [review]
Current state looks good! Once Jenkins is done I'd like another engineer to review as well... Please assign over to either @ejohnstown or @cconlon for final. Thanks @lealem47

**Review by douzzer — CHANGES_REQUESTED** [review]
There are some significant issues here.

First, we need to rethink how to streamline the `client` and `server` in the examples directory.  With the modest culling here, the line count is not meaningfully lower, test_app vs examples.

In other words, what we're mainly accomplishing here is forking our codebase, then maintaining two very complicated nearly identical versions of the example client and server.  That's not viable, in my estimation.

The danger of doing this is already apparent in the PR, in that the test_app client and server are already out of date with master and missing important functionality added since the fork.

In my estimation, there are two ways we can move forward with this: either the client and server in examples/ need to be refactored much more extensively, or the benchmark and other testing-specific code in the full test_app versions need to be moved into separate files that are conditionally linked to create the test_app executables, but there is only one client.c and one server.c, with short spans of gated code in them to conditionally include the testing functionality.

I think the latter plan is probably the way to go, at least for now.

I do think it's right to move client.c and server.c to a new test_app directory.

If we do indeed move forward with that, the relocation needs to be redone with a `git mv`, with any content changes afterwards, to retain the version history and eliminate the possibility of lost changes during the merge process.

And if the example client and server do wind up as completely separate applications, their source files should have new names, e.g. example_{client,server}.{c,h}.

**Comment by lealem47** [comment]
Idle.
